keps/sig-node/20190221-initializationFailureThreshold-health-probes.md

```diff
@@ -0,0 +1,149 @@
+  - Clearly state that PostStart handlers do not delay probe executions.
+  - Introduce and explain this new option.
+  - Document that `kubelet` does not save states, and what are the implications with this new option (see Risks and Mitigations).
+  - Document appropriate use cases for this new option.
+
+### Non-Goals
+
+- This proposal does not address the issue of pod load affecting startup (or any other probe that may be delayed due to load). It is acting strictly at the pod level, not the node level.
+- This proposal will only update the official Kubernetes documentation, excluding [A Pod's Life] and other well referenced pages explaining probes.
+
+[A Pod's Life]: https://blog.openshift.com/kubernetes-pods-life/
+
+## Proposal
+
+### Implementation Details
+
+The proposed solution is to add a new `int32` field named `InitializationFailureThreshold` to the type `Probe` of the core API, which is mapped to the numerical option `initializationFailureThreshold`.
+
+It also requires keeping the state of the container (has the probe ever succeeded?) using a boolean `hasInitialized` inside the kubelet worker.
+
+The combination of `hasInitialized` false and `resultRun` count lower than `initializationFailureThreshold` becomes another condition to return `true` to the probe state in `worker.go`.
+
+For example, if `periodSeconds` is 10, `initializationFailureThreshold` is 20, and `failureThreshold` is 3, it means that:
+
+- The kubelet will allow the container 200 seconds to start (20 probes, spaced 10 seconds apart).
+- If a probe succeeds at any time during that interval, the container is considered to have started, and `failureThreshold` is used thereafter.
+
+This means that all these cases will lead to a container being terminated:
+
+- The container fails 20 probes at startup. It is considered to have failed, and is terminated after 200 seconds of downtime.
+- The container fails 10 probes at startup, starts successfully, and after a long time fails 3 probes. The container is considered to have failed and is terminated after 30 seconds of downtime.
+- The container fails 10 probes at startup, succeeds once, and fails 3 more probes. The container is considered to have started at 100 seconds, and even though it is still within the first 200 seconds of its lifetime covered by `initializationFailureThreshold`, it is considered to have failed because of `failureThreshold` and the fact that it had initially started successfully, therefore it is terminated after 30 seconds of downtime.
+
+If `initializationFailureThreshold` is set smaller than `failureThreshold`, it's value is overridden to `failureThreshold` to avoid having the container being killed faster during startup than it would be in case of a deadlock, rendering is permanently unavailable.
+
+This is being implemented and reviewed in PR [#71449].
+
+[#71449]: https://github.com/kubernetes/kubernetes/pull/71449
+
+### Risks and Mitigations
+
+#### Stateless kubelet
```

Comment:
the kep should include a section on graduation criteria?

i assume this new field would have to go through the normal process of feature gates.

Reply:
There is already an API review here: https://github.com/orgs/kubernetes/projects/13#card-19727773
I am quite new to the KEP process, do you mean something else?